### PR TITLE
mgr/dashboard: cluster upgrade base UI component 

### PR DESCRIPTION
**Pull request by nizamial09**
Follows #52222

https://github.com/ceph/ceph/assets/71764184/6fe41a25-e29b-4758-8d9f-4d8d236ad18c

**When the cluster is already updated**
image

- Base layout for the Upgrade Component
- Shows the Upgrade available button when an upgrade is available

Fixes: https://tracker.ceph.com/issues/61927
Signed-off-by: Nizamudeen A <nia@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by cloudbehl** [comment]
Also, make sure the PR is properly divided and not bigger to review.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade.component.html [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                i18n>
+          <i [ngClass]="[icons.up]"
+             class="text-info"></i>
+            Upgrade available</span>
+          <div i18n-ngbTooltip
+               [ngbTooltip]="(healthData.mgr_map | mgrSummary).total <= 1 ? 'To upgrade, you need minimum 2 mgr daemons.' : ''">
+            <button class="btn btn-accent mt-2"
+                    id="upgrade"
+                    aria-label="Upgrade now"
+                    [disabled]="(healthData.mgr_map | mgrSummary).total <= 1"
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade.component.html:42]: This query seems costly just to get count of the mgr, isn't it? Maybe we can get it from orch api, wdyt?

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/upgrade.service.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    return this.http.get(this.baseURL).pipe(
+      map((resp: UpgradeInfoInterface) => {
+        return this.versionAvailableForUpgrades(resp);
+      })
+    );
+  }
+
+  versionAvailableForUpgrades(upgradeInfo: UpgradeInfoInterface): UpgradeInfoInterface {
+    this.summaryService.subscribe((summary) => {
+      const version = summary.version.replace('ceph version ', '').split('-');
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/api/upgrade.service.ts:29]: Maybe this code snippet can be made simpler by directly filtering `upgradeInfo.versions`?
```suggestion
const versionNumber = summary.version.replace('ceph version ', '').split('-')[0];
const upgradableVersions = upgradeInfo.versions.filter(targetVersion => {
  const cVersion = versionNumber.split('.');
  const tVersion = targetVersion.split('.');

  return cVersion[0] === tVersion[0] && (cVersion[1] < tVersion[1] || cVersion[2] < tVersion[2]);
});

upgradeInfo.versions = upgradableVersions;
return upgradeInfo; 
```

**Review by avanthakkar — APPROVED** [review]
Tested locally and LGTM! Just left few code improvements.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/upgrade.service.ts (hunk 2 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/api/upgrade.service.ts]: thanks @avanthakkar updated!

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade.component.html (hunk 1 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade.component.html:42]: we are also displaying the health status and to get the health status i'll anyway need to get the healthData. So its less costlier than adding another api call.

**Review by aaSharma14 — APPROVED** [review]

**Comment by avanthakkar** [comment]
Does the upgrade works if we have the ceph health in WARN state? Let's say we have `TOO_FEW_OSDS` warn state, does upgrade throws error or it starts upgrade for other ceph daemons? @nizamial09

**Comment by nizamial09** [comment]
> Does the upgrade works if we have the ceph health in WARN state? Let's say we have `TOO_FEW_OSDS` warn state, does upgrade throws error or it starts upgrade for other ceph daemons? @nizamial09

Probably @adk3798 could answer this better but my guess is the upgrade will start and proceed regardless of the Cluster WARN_STATE.

**Comment by cloudbehl** [comment]
jenkins test make check

**Pull request merged**